### PR TITLE
Messaging refactor develop

### PR DESCRIPTION
**Pull request by hongdami**
Issue Number: 

Objective of pull request:

## Pull request checklist

Your PR fulfills the following requirements:
- [ ] [Issue](https://github.com/lava-nc/lava/issues) created that explains the change and why it's needed
- [x] Tests are part of the PR (for bug fixes / features)
- [ ] [Docs](https://github.com/lava-nc/docs) reviewed and added / updated if needed (for bug fixes / features)
- [ ] PR conforms to [Coding Conventions](https://lava-nc.org/developer_guide.html#coding-conventions)
- [ ] [PR applys BSD 3-clause or LGPL2.1+ Licenses](https://lava-nc.org/developer_guide.html#add-a-license) to all code files
- [x] Lint (`flakeheaven lint src/lava tests/`) and (`bandit -r src/lava/.`) pass locally
- [ ] Build tests (`pytest`) passes locally

## Pull request type

 

 

Please check your PR type:

- [x] Bugfix
- [x] Feature
- [x] Code style update (formatting, renaming)
- [ ] Refactoring (no functional changes, no api changes)
- [ ] Build related changes
- [ ] Documentation changes
- [ ] Other (please describe): 

## What is the current behavior?

-

## What is the new behavior?

-

## Does this introduce a breaking change?

- [ ] Yes
- [ ] No

### CODE REVIEW

File: src/lava/magma/runtime/message_infrastructure/message_infrastructure/multiprocessing.py [hunk 1]
```diff
@@ -0,0 +1,99 @@
...
+    PARENT_PROC = 0
+    CHILD_PROC = 1
+
+
+class MultiProcessing(MessageInfrastructureInterface):
+    """Implements message passing using shared memory and multiprocessing"""
+
+    def __init__(self):
+        self._mp: ty.Optional[CppMultiProcessing] = None
+        self._smm: ty.Optional[SharedMemoryManager] = None
```
Comment (killight98) [inline src/lava/magma/runtime/message_infrastructure/message_infrastructure/multiprocessing.py:53]: replace `SharedMemoryManager` by `SharedMemManager`.

File: src/lava/magma/runtime/message_infrastructure/message_infrastructure/multiprocessing.py [hunk 2]
```diff
@@ -0,0 +1,99 @@
...
+
+    @property
+    def smm(self):
+        """Returns the underlying shared memory manager"""
+        return self._smm
+
+    def start(self):
+        """Starts the shared memory manager"""
+        self._mp = CppMultiProcessing()
+        self._smm = SharedMemoryManager()
```
Comment (killight98) [inline src/lava/magma/runtime/message_infrastructure/message_infrastructure/multiprocessing.py:69]: same with before.

**Review by killight98 — CHANGES_REQUESTED** [review]

File: src/lava/magma/runtime/message_infrastructure/message_infrastructure/multiprocessing.py [hunk 3]
```diff
@@ -0,0 +1,99 @@
...
+            SyncDomain, 'RuntimeServiceBuilder']]) -> ty.Any:
+        """Given a target_fn starts a system (os) process"""
+
+        ret = self._mp.build_actor()
+        if ret == ProcessType.ERR_PROC:
+            exit(-1)
+        if ret == ProcessType.CHILD_PROC:
+            kwargs = {"builder": builder}
+            target_fn(**kwargs)
+            exit(0)
```
Comment (killight98) [inline src/lava/magma/runtime/message_infrastructure/message_infrastructure/multiprocessing.py:82]: i think MultiProcessing/manager should handle the exit/stop.

**Review by killight98 — CHANGES_REQUESTED** [review]

File: src/lava/magma/runtime/message_infrastructure/message_infrastructure/multiprocessing.py [hunk 4]
```diff
@@ -0,0 +1,99 @@
...
+
+
+"""Implements the Message Infrastructure Interface using Python
+MultiProcessing Library. The MultiProcessing API is used to create actors
+which will participate in exchanging messages. The Channel Infrastructure
+further uses the SharedMemoryManager from MultiProcessing Library to
+implement the communication backend in this implementation."""
+
+
+class ProcessType(Enum):
```
Comment (killight98) [inline src/lava/magma/runtime/message_infrastructure/message_infrastructure/multiprocessing.py:42]: mv this enum to pythonwrapper and bind the `ProcessType`

File: src/lava/magma/runtime/message_infrastructure/message_infrastructure/csrc/multiprocessing.h [hunk 5]
```diff
@@ -13,11 +13,17 @@
 
 namespace message_infrastructure {
 
+enum ChannelType {
```
Comment (killight98) [inline src/lava/magma/runtime/message_infrastructure/message_infrastructure/csrc/multiprocessing.h:16]: rename as `ProcessType`？

**Review by killight98 — CHANGES_REQUESTED** [review]

File: src/lava/magma/runtime/message_infrastructure/message_infrastructure/multiprocessing.py [hunk 6]
```diff
@@ -0,0 +1,99 @@
...
+class MultiProcessing(MessageInfrastructureInterface):
+    """Implements message passing using shared memory and multiprocessing"""
+
+    def __init__(self):
+        self._mp: ty.Optional[CppMultiProcessing] = None
+        self._smm: ty.Optional[SharedMemoryManager] = None
+        self._actors: ty.List[Actor] = []
+
+    @property
+    def actors(self):
```
Comment (killight98) [inline src/lava/magma/runtime/message_infrastructure/message_infrastructure/multiprocessing.py:57]: get actors from actors_ at cpp

File: src/lava/magma/runtime/message_infrastructure/message_infrastructure/multiprocessing.py [hunk 7]
```diff
@@ -0,0 +1,99 @@
...
+    CHILD_PROC = 1
+
+
+class MultiProcessing(MessageInfrastructureInterface):
+    """Implements message passing using shared memory and multiprocessing"""
+
+    def __init__(self):
+        self._mp: ty.Optional[CppMultiProcessing] = None
+        self._smm: ty.Optional[SharedMemoryManager] = None
+        self._actors: ty.List[Actor] = []
```
Comment (killight98) [inline src/lava/magma/runtime/message_infrastructure/message_infrastructure/multiprocessing.py:54]: remove it.

File: src/lava/magma/runtime/message_infrastructure/message_infrastructure/multiprocessing.py [hunk 8]
```diff
@@ -0,0 +1,99 @@
...
+
+    @property
+    def actors(self):
+        """Returns a list of actors"""
+        return self._actors
+
+    @property
+    def smm(self):
```
Comment (killight98) [inline src/lava/magma/runtime/message_infrastructure/message_infrastructure/multiprocessing.py:62]: same as `actors` func

File: src/lava/magma/runtime/message_infrastructure/message_infrastructure/multiprocessing.py [hunk 9]
```diff
@@ -0,0 +1,99 @@
...
+        """Returns the underlying shared memory manager"""
+        return self._smm
+
+    def start(self):
+        """Starts the shared memory manager"""
+        self._mp = CppMultiProcessing()
```
Comment (killight98) [inline src/lava/magma/runtime/message_infrastructure/message_infrastructure/multiprocessing.py:68]: move to `__init__`

**Review by killight98 — CHANGES_REQUESTED** [review]

File: src/lava/magma/compiler/builders/channel_builder.py [hunk 10]
```diff
@@ -15,7 +15,7 @@
     ChannelType,
 )
 from lava.magma.compiler.utils import PortInitializer
-from lava.magma.runtime.message_infrastructure \
+from lava.magma.runtime.message_infrastructure.message_infrastructure \
```
Comment (killight98) [inline src/lava/magma/compiler/builders/channel_builder.py:18]: from message_infrastructure.message_infrastructure_interface import (MessageInfrastructureInterface)

File: src/lava/magma/compiler/channels/pypychannel.py [hunk 11]
```diff
@@ -17,7 +17,8 @@
 
 if ty.TYPE_CHECKING:
     from lava.magma.runtime.message_infrastructure \
-        .message_infrastructure_interface import (
+        .message_infrastructure.message_infrastructure_interface \
```
Comment (killight98) [inline src/lava/magma/compiler/channels/pypychannel.py:20]: same as before

File: src/lava/magma/runtime/runtime.py [hunk 12]
```diff
@@ -13,10 +13,10 @@
 from lava.magma.compiler.channels.pypychannel import CspRecvPort, CspSendPort
 from lava.magma.compiler.var_model import AbstractVarModel
 from lava.magma.core.process.message_interface_enum import ActorType
-from lava.magma.runtime.message_infrastructure.factory import \
-    MessageInfrastructureFactory
-from lava.magma.runtime.message_infrastructure. \
-    message_infrastructure_interface import \
+from lava.magma.runtime.message_infrastructure.message_infrastructure \
+    .factory import MessageInfrastructureFactory
+from lava.magma.runtime.message_infrastructure.message_infrastructure \
```
Comment (killight98) [inline src/lava/magma/runtime/runtime.py:18]: same as before

**Review by killight98 — CHANGES_REQUESTED** [review]

**Review by hongdami — COMMENTED** [review]
1. update import path and __init__.py
2. fix typoerror
3. update bounded function to cpp
4. update pywrapper
5. add test for multiprocessing.py

File: src/lava/magma/runtime/message_infrastructure/message_infrastructure/factory.py [hunk 13]
```diff
@@ -0,0 +1,19 @@
+# Copyright (C) 2021-22 Intel Corporation
+# SPDX-License-Identifier: LGPL 2.1 or later
+# See: https://spdx.org/licenses/
+from lava.magma.core.process.message_interface_enum import ActorType
```
Comment (killight98) [inline src/lava/magma/runtime/message_infrastructure/message_infrastructure/factory.py:4]: porting ActorType

File: src/lava/magma/runtime/message_infrastructure/message_infrastructure/factory.py [hunk 14]
```diff
@@ -0,0 +1,19 @@
...
+from lava.magma.runtime.message_infrastructure.message_infrastructure \
+    .multiprocessing import MultiProcessing
```
Comment (killight98) [inline src/lava/magma/runtime/message_infrastructure/message_infrastructure/factory.py:6]: change import

**Review by killight98 — CHANGES_REQUESTED** [review]

**Review by killight98 — APPROVED** [review]

**Review by hexu33 — APPROVED** [review]

**Pull request merged**